src/relax/transform/remove_unused_parameters.cc

```diff
@@ -0,0 +1,239 @@
+        new_args.push_back(old_args[i]);
+      }
+    }
+
+    if (free_tir_vars.size()) {
+      Map<Var, Expr> old_binding;
+      for (size_t i = 0; i < old_relax_params.size(); i++) {
+        old_binding.Set(old_relax_params[i], old_args[i]);
+      }
+      arith::Analyzer analyzer;
+      auto tir_binding = InferSymbolicVarMap(old_binding, &analyzer);
+
+      for (const auto& tir_var : free_tir_vars) {
+        new_args.push_back(PrimValue(tir_binding.at(tir_var)));
+      }
+    }
+
+    return new_args;
+  };
+
+  auto write_ptr = func.CopyOnWrite();
+  write_ptr->params = params;
+  write_ptr->struct_info_ = new_sinfo;
+
+  return CalleeAnalysis{func, arg_updater};
+}
+
+class CallSiteMutator : public ExprMutator {
+ public:
+  explicit CallSiteMutator(PMap<GlobalVar, std::function<Call(Call)>> callsite_updaters)
+      : callsite_updaters_(callsite_updaters) {}
+
+  using ExprMutator::VisitExpr_;
+
+  Expr VisitExpr_(const CallNode* op) override {
+    auto node = Downcast<Call>(ExprMutator::VisitExpr_(op));
+
+    if (auto gvar = node->op.as<GlobalVar>()) {
+      if (auto it = callsite_updaters_.find(gvar.value()); it != callsite_updaters_.end()) {
+        node = it->second(std::move(node));
+      }
+    }
+
+    return node;
+  }
+
+  PMap<GlobalVar, std::function<Call(Call)>> callsite_updaters_;
+};
+
+}  // namespace
+
+namespace transform {
+
+Pass RemoveUnusedParameters() {
+  runtime::TypedPackedFunc<IRModule(IRModule, PassContext)> pass_func =
+      [=](IRModule mod, PassContext pc) -> IRModule {
+    PMap<GlobalVar, std::function<Call(Call)>> callsite_updaters;
+
+    {
+      IRModule new_callees;
+
+      for (const auto& [gvar, base_func] : mod->functions) {
+        if (auto func = base_func.as<Function>()) {
+          if (auto callee_res = AnalyzeCallee(func.value())) {
+            auto new_func = callee_res->func;
+            GlobalVar new_gvar(gvar->name_hint, new_func->checked_type_);
+            new_gvar->struct_info_ = new_func->struct_info_;
+            new_callees->Add(new_gvar, new_func);
+
+            callsite_updaters[gvar] = [old_gvar = gvar, new_gvar,
+                                       arg_updater = callee_res->arg_updater](Call call) -> Call {
+              ICHECK(call->op.same_as(old_gvar)) << "InternalError: "
+                                                 << "Updater should be applied to " << old_gvar
+                                                 << ", but was applied to " << call->op;
+              auto write_ptr = call.CopyOnWrite();
+              write_ptr->op = new_gvar;
+              write_ptr->args = arg_updater(call->args);
+              return call;
+            };
+          }
+        }
+      }
+
+      if (callsite_updaters.empty()) {
+        return mod;
+      }
+      auto write_ptr = mod.CopyOnWrite();
+      for (const auto& it : callsite_updaters) {
+        write_ptr->Remove(it.first);
+      }
+      write_ptr->Update(new_callees);
```

Comment:
Might be worth putting in a comment that this is to transfer over the functions that will be remaining unchanged. I had to think for a second to realize what this was doing.

Reply:
Good call, and I've I added a comment.  This section is to remove any private functions that have unused parameters, then to replace them with updated versions of the private functions.